python/ql/lib/semmle/python/security/dataflow/TarSlipCustomizations.qll

```diff
@@ -55,10 +55,48 @@ module TarSlip {
     ExcludeTarFilePy() { this.getLocation().getFile().getBaseName() = "tarfile.py" }
   }
 
+  private DataFlow::TypeTrackingNode unsafeFilter(DataFlow::TypeTracker t) {
+    t.start() and
+    (
+      result.asExpr().(StrConst).getS() = "fully_trusted"
+      or
+      result.asExpr() instanceof None
+    )
+    or
+    exists(DataFlow::TypeTracker t2 | result = unsafeFilter(t2).track(t2, t))
+  }
+
+  private DataFlow::Node unsafeFilter() {
+    unsafeFilter(DataFlow::TypeTracker::end()).flowsTo(result)
+  }
+
+  /**
+   * Holds if `call` has an unsafe extraction filter, either by default (as the default is unsafe),
+   * or by being set to an explicitly unsafe value, such as `"fully_trusted"`, or `None`.
+   */
+  private predicate hasUnsafeFilter(DataFlow::CallCfgNode call) {
+    call =
+      API::moduleImport("tarfile")
+          .getMember("open")
+          .getReturn()
+          .getMember(["extract", "extractall"])
+          .getACall() and
+    (
+      call.getArg(4) = unsafeFilter()
+      or
+      call.getArgByName("filter") = unsafeFilter()
+      or
+      not exists(call.getArg(4)) and not exists(call.getArgByName("filter"))
+    )
+  }
```

Comment:
If we use `API::CallNode` we can entirely get rid of the manual type-tracker. Should be possible to convert all the `DataFlow::CallCfgNode` used in this file to `API::CallNode` as well.

```suggestion
  private predicate hasUnsafeFilter(API::CallNode call) {
    call =
      API::moduleImport("tarfile")
          .getMember("open")
          .getReturn()
          .getMember(["extract", "extractall"])
          .getACall() and
    (
      exists(Expr filterValue |
        filterValue = call.getParameter(4, "filter").getAValueReachingSink().asExpr() and
        (
          filterValue.(StrConst).getText() = "fully_trusted"
          or
          filterValue instanceof None
        )
      )
      or
      not exists(call.getParameter(4, "filter"))
    )
  }
```

Reply:
Oh that is so much nicer! ❤️